### PR TITLE
OkHttpClients uses ExponentialBackoff instead of NoDelayBackoff

### PR DESCRIPTION
**Pull request by uschi2000**

### CODE REVIEW

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java [hunk 1]
```diff
@@ -16,23 +16,34 @@
...
-        return numTimesRetried <= maxNumRetries
-                ? Optional.of(Duration.ZERO)
-                : Optional.empty();
+        retryNumber += 1;
+        if (retryNumber > maxNumRetries) {
+            return Optional.empty();
+        }
+
+        int upperBound = (int) Math.pow(2, retryNumber);
+        return Optional.of(backoffSlotSize.multipliedBy(random.nextInt(upperBound)));
```
Comment (schlosna) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java:51]: do we want [decorrelated jitter](https://www.awsarchitectureblog.com/2015/03/backoff.html)?

Comment (bavardage) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java:51]: > Of the jittered approaches, “Equal Jitter” is the loser. It does slightly more work than “Full Jitter”, and takes much longer. The decision between “Decorrelated Jitter” and “Full Jitter” is less clear. The “Full Jitter” approach uses less work, but slightly more time. Both approaches, though, present a substantial decrease in client work and server load.

so this, as implemented, is full jitter, right?

Comment (schlosna) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java:51]: roughly yes. mostly just want us to define/document the intended algorithm. I think in most cases we're probably happy to do more work if that reduces time.

Comment (uschi2000) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java:51]: i think in the grander scheme of things, it probably doesn't matter for today. i added some docs.

**Review by bavardage — APPROVED** [review]

**Pull request merged**